providers/rfc2136.go

```diff
@@ -0,0 +1,182 @@
+	port       string
+	keyname    string
+	key        string
+	rootdomain string
+}
+
+/* Private functions */
+func needFQDN(domain string) string {
+	if (strings.HasSuffix(domain, ".")) {
+		return domain
+	} else {
+		return domain + "."
+	}
+}
+
+func (b RFC2136Handler) sendMessage(msg *dns.Msg) error {
+	t := new(dns.Transfer)
+
+	nameserver := net.JoinHostPort(b.server, b.port)
+	msg.SetTsig(b.keyname, "hmac-md5.sig-alg.reg.int.", 300, time.Now().Unix())
+	t.TsigSecret = map[string]string{b.keyname: b.key}
+
+	_, err := t.In(msg, nameserver)
+	t.Close()
+	return err
+}
+
+func (b RFC2136Handler) add(fqdn string, ip string, class string, ttl int) error {
+	logrus.Debugf("adding dns entry %s %s %d", fqdn, ip, ttl)
+	m := new(dns.Msg)
+	m.SetUpdate(b.rootdomain)
+	rr, err := dns.NewRR(fmt.Sprintf("%s %d %s %s", needFQDN(fqdn), ttl, class, ip))
+	if (err != nil) {
+		return err
+	}
+	rrs := make([]dns.RR, 1)
+	rrs[0] = rr
+	m.Insert(rrs)
+	return b.sendMessage(m)
+}
+
+func (b RFC2136Handler) remove(fqdn string, class string) error {
+	logrus.Debugf("removing dns entry %s", fqdn)
+	m := new(dns.Msg)
+	m.SetUpdate(b.rootdomain)
+	rr, rrerr := dns.NewRR(fmt.Sprintf("%s 0 %s 0.0.0.0", needFQDN(fqdn), class))
+	if (rrerr != nil) {
+		return rrerr
+	}
+	rr.Header().Class = dns.TypeANY
+	rrs := make([]dns.RR, 1)
+	rrs[0] = rr
+	m.RemoveRRset(rrs)
+	return b.sendMessage(m)
+}
+
+func (b RFC2136Handler) list() ([]dns.RR, error) {
+	logrus.Debugf("listing dns entries for %s", b.rootdomain)
+	t := new(dns.Transfer)
+	t.TsigSecret = map[string]string{b.keyname: b.key}
+
+	m := new(dns.Msg)
+	m.Question = make([]dns.Question, 1)
+
+	nameserver := net.JoinHostPort(b.server, b.port)
+	m.SetAxfr(b.rootdomain)
+
+	env, err := t.In(m, nameserver)
+	if err != nil {
+		return nil, err;
+	}
+	envelope := 0
+	record := 0
+	records := make([]dns.RR, 0)
+
+	// TODO what is an enveloppe, can i receive several of them ?
+	for e := range env {
+		if e.Error != nil {
+			fmt.Printf(";; %s\n", e.Error.Error())
+			return nil, err
+		}
+		records = append(records, e.RR...)
+		record += len(e.RR)
+		envelope++
+	}
+	t.Close()
+	return records, nil
+}
+
+func init() {
+	server := os.Getenv("DNS_HOST")
+	port := os.Getenv("DNS_PORT")
+	keyname := os.Getenv("TSIG_KEYNAME")
+	key := os.Getenv("TSIG_KEY")
+	if len(server) == 0 || len(port) == 0 || len(keyname) == 0 || len(key) == 0 {
+		logrus.Info("RFC2136 environnement not set, skipping init of RFC2136 provider")
+		return
+	}
+
+	rfc2136Handler := &RFC2136Handler{server, port, needFQDN(keyname), key, needFQDN(rdns.RootDomainName)}
+	if err := RegisterProvider("RFC2136", rfc2136Handler); err != nil {
+		logrus.Fatal("Could not register RFC2136 provider")
+	}
+	logrus.Infof("Configured %s with zone %s and server %s", rfc2136Handler.GetName(), rdns.RootDomainName, server)
+
+}
+
+/* Provider implementation */
+func (b *RFC2136Handler) AddRecord(record rdns.DnsRecord) error {
+	for _, rec := range record.Records {
```

Comment:
Instead of making multiple calls to the nameserver for each record you should use a single call passing all records at once.